### PR TITLE
Improve birthdate validation

### PR DESCRIPTION
**Pull request by zessx**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Birthdate validation use some useless code, which could be improved
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | -
| How to test?  | See following code

```php
$tests = array(
	array(true, '1991-04-19'),
	array(true, '2015-03-22'),
	array(true, '1945-07-25'),
	array(false, '2020-03-19'),
	array(false, '1991-03-33'), 
	array(false, '1991-15-19'),
	array(true, date('Y-m-d', strtotime('now'))), 
	array(true, date('Y-m-d', strtotime('-1 day'))),
	array(false, date('Y-m-d', strtotime('+1 day'))),
	array(true, date('Y-m-d', strtotime('-1 month'))),
	array(true, date('Y-m-d', strtotime('-1 month -1 day'))),
	array(true, date('Y-m-d', strtotime('-1 month +1 day'))),
	array(false, date('Y-m-d', strtotime('+1 month'))),
	array(false, date('Y-m-d', strtotime('+1 month -1 day'))),
	array(false, date('Y-m-d', strtotime('+1 month +1 day'))),
	array(true, date('Y-m-d', strtotime('-1 year'))),
	array(true, date('Y-m-d', strtotime('-1 year -1 day'))),
	array(true, date('Y-m-d', strtotime('-1 year +1 day'))),
	array(true, date('Y-m-d', strtotime('-1 year -1 month'))),
	array(true, date('Y-m-d', strtotime('-1 year -1 month -1 day'))),
	array(true, date('Y-m-d', strtotime('-1 year -1 month +1 day'))),
	array(true, date('Y-m-d', strtotime('-1 year +1 month'))),
	array(true, date('Y-m-d', strtotime('-1 year +1 month -1 day'))),
	array(true, date('Y-m-d', strtotime('-1 year +1 month +1 day'))),
	array(false, date('Y-m-d', strtotime('+1 year'))),
	array(false, date('Y-m-d', strtotime('+1 year -1 day'))),
	array(false, date('Y-m-d', strtotime('+1 year +1 day'))),
	array(false, date('Y-m-d', strtotime('+1 year -1 month'))),
	array(false, date('Y-m-d', strtotime('+1 year -1 month -1 day'))),
	array(false, date('Y-m-d', strtotime('+1 year -1 month +1 day'))),
	array(false, date('Y-m-d', strtotime('+1 year +1 month'))),
	array(false, date('Y-m-d', strtotime('+1 year +1 month -1 day'))),
	array(false, date('Y-m-d', strtotime('+1 year +1 month +1 day'))),
);

foreach ($tests as $test) {
	$date = $test[1];
	$expected = $test[0];
	
	$isBirthDate = Validate::isBirthDate($date);
	$success = $isBirthDate === $expected;
	printf(
		'[%s]: is%s a valid birthdate <span style="color:%s">(test %s)</span><br>',
		$date,
		$isBirthDate ? '' : ' not',
		$success ? 'forestgreen' : 'tomato',
		$success ? 'successful' : 'failed'
	);
}
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11470)

### CODE REVIEW

File: classes/Validate.php [hunk 1]
```diff
@@ -602,15 +602,7 @@ public static function isBirthDate($date)
...
-            if ($birth_date[1] > date('Y')
-                || ($birth_date[1] > date('Y') && $birth_date[2] > date('m'))
-                || ($birth_date[1] > date('Y') && $birth_date[2] > date('m') && $birth_date[3] > date('d'))
-                || ($birth_date[1] == date('Y') && $birth_date[2] == date('m') && $birth_date[3] > date('d'))
-                || ($birth_date[1] == date('Y') && $birth_date[2] > date('m'))) {
-                return false;
-            }
-
-            return true;
+            return strtotime($date) < time();
```
Comment (PierreRambaud) [inline classes/Validate.php]: There is a problem with strtotime, it only parse any English textual datetime. This is why there is this complex regex ^^

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
You must check with other date format.

File: classes/Validate.php (hunk 1 above)
Comment (zessx) [reply classes/Validate.php]: At this point, we already passed the regex test, to ensure we're using an English-format date. Therefore there is no problem to use the `$date` variable.

Previous code was already assuming this, trusing the regex test:
```php
$birth_date[2] > date('m')
```

Comment (PierreRambaud) [reply classes/Validate.php]: Woot my bad, miss reading the regex :/

But it should be better to use something like:
```php
$d = DateTime::createFromFormat($date, 'Y-m-d');
if (false === $d) {
  return false;
}
return $d->getTimestamp() < time();
```
What do you think?

Comment (zessx) [reply classes/Validate.php]: I totally agree with this!

Comment (PierreRambaud) [reply classes/Validate.php]: Update your pull request, and we will be glad to merge it! :)
And thanks for your contribution!

**Comment by zessx** [comment]
Sorry, I've been busy and totally forgot this PR.  
Updated, with `DateTime` usage.

**Review by PierreRambaud — DISMISSED** [review]

**Comment by mbadrani** [comment]
@PierreRambaud do you have an idea on how to test this except looking at the code?

**Comment by PierreRambaud** [comment]
Create an account and make sure all tests that @zessx created are still valid :)
Example:
-1 year -1 day => 2018/01/22 (if we are the 2019/01/23)
+1 year +1 month +1 day => 2020/02/24

**Comment by Quetzacoalt91** [comment]
The provided cases could be added in the existing tests:

```php
    public function isBirthDateProvider()
    {
        return array(
            array(true, '1991-04-19'),
            array(true, '2015-03-22'),
            array(true, '1945-07-25'),
            array(false, '2020-03-19'),
            array(false, '1991-03-33'), // <- nope
            array(false, '1991-15-19'),
            array(true, date('Y-m-d', strtotime('now'))), 
            array(true, date('Y-m-d', strtotime('-1 day'))),
            array(false, date('Y-m-d', strtotime('+1 day'))),
            array(true, date('Y-m-d', strtotime('-1 month'))),
            array(true, date('Y-m-d', strtotime('-1 month -1 day'))),
            array(true, date('Y-m-d', strtotime('-1 month +1 day'))),
            array(false, date('Y-m-d', strtotime('+1 month'))),
            array(false, date('Y-m-d', strtotime('+1 month -1 day'))),
            array(false, date('Y-m-d', strtotime('+1 month +1 day'))),
            array(true, date('Y-m-d', strtotime('-1 year'))),
            array(true, date('Y-m-d', strtotime('-1 year -1 day'))),
            array(true, date('Y-m-d', strtotime('-1 year +1 day'))),
            array(true, date('Y-m-d', strtotime('-1 year -1 month'))),
            array(true, date('Y-m-d', strtotime('-1 year -1 month -1 day'))),
            array(true, date('Y-m-d', strtotime('-1 year -1 month +1 day'))),
            array(true, date('Y-m-d', strtotime('-1 year +1 month'))),
            array(true, date('Y-m-d', strtotime('-1 year +1 month -1 day'))),
            array(true, date('Y-m-d', strtotime('-1 year +1 month +1 day'))),
            array(false, date('Y-m-d', strtotime('+1 year'))),
            array(false, date('Y-m-d', strtotime('+1 year -1 day'))),
            array(false, date('Y-m-d', strtotime('+1 year +1 day'))),
            array(false, date('Y-m-d', strtotime('+1 year -1 month'))),
            array(false, date('Y-m-d', strtotime('+1 year -1 month -1 day'))),
            array(false, date('Y-m-d', strtotime('+1 year -1 month +1 day'))),
            array(false, date('Y-m-d', strtotime('+1 year +1 month'))),
            array(false, date('Y-m-d', strtotime('+1 year +1 month -1 day'))),
            array(false, date('Y-m-d', strtotime('+1 year +1 month +1 day'))),
        );
    }
```

But did anybody saw the existing tests are now broken because of the changes? It seems some check have been removed.

**Comment by zessx** [comment]
@Quetzacoalt91 Thanks for your message. The use of `DateTime` indeed remove the regex check, which was ensuring we're not using fake dates as `1991-03-33`.

I revert to the previous commit, which passes all the tests.

**Comment by PierreRambaud** [comment]
@zessx @Quetzacoalt91 Tests were failing because of another issue (Addons api). The datetime approach is better because of strtotime can change with timezone :/

**Comment by Quetzacoalt91** [comment]
@PierreRambaud, in https://travis-ci.org/PrestaShop/PrestaShop/jobs/483444309 we can see the actual error:

```
There was 1 failure:
1) LegacyTests\Unit\Classes\ValidateCoreTest::testIsBirthDate with data set #4 (false, '1991-03-33')
Failed asserting that true matches expected false.
/home/travis/build/PrestaShop/PrestaShop/tests-legacy/Unit/Classes/ValidateCoreTest.php:66
```

The latest commit seems to fix this regression, and I allowed myself to add the cases given by this PR.

**Comment by PierreRambaud** [comment]
Holy f**k, I'm a little bit tired --'.

This is the correct way to test if a date is valid :smile: 

```php
$d = DateTime::createFromFormat($date, 'Y-m-d');
if (!empty(DateTime::getLastErrors()['warning_count']) || false === $d) {
  return false;
}

return $d->getTimestamp() < time();
```

**Comment by Quetzacoalt91** [comment]
Looks like the codebase does definitely not want your check with `Datetime`, all tests are broken. :)

**Comment by PierreRambaud** [comment]
Oops, should be ok now.

**Review by Quetzacoalt91 — DISMISSED** [review]

**Comment by marionf** [comment]
Hello @zessx 

I can create an account with these dates: 
1991-03-33, in the customer account 1991-04-02 is saved
1991-15-19, in the customer account 1992-03-19 is saved

https://drive.google.com/file/d/12TV1qGlxS4dioaM493dyC59rXlBBvOJt/view?usp=sharing

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @zessx

**Pull request merged**